### ISSUE TITLE
PnP-JS Code not working when embedded on Content Editor WP on IE 11 browser

### ISSUE THREAD

**Issue by astatke** [SharePoint/PnP-JS-Core#432]
[x] Bug

[x] Question

Please specify what version of the library you are using: [  sp-pnp-js v2.0.4 ]

#### Expected / Desired Behavior 
I expected the Title of my SP 2016 on-premise site be displayed on IE browser. I have no issues on  Chrome and Firefox. I was basically trying to replicate the example you have on this github page but to no avail, https://github.com/SharePoint/PnP-JS-Core/wiki/Getting-Started:-Install-&-Use.

#### Observed Behavior
I see that required Polyfill files are loaded but the code is giving this error i.e. as seen in the attached screenshot file, SCRIPT5009: 'Promise' is undefined. File: pnp.js, Line: 355, Column: 9. 

#### Steps to Reproduce
Follow the steps as outlined under the subtitle "From a Content Editor WebPart" from this github page, https://github.com/SharePoint/PnP-JS-Core/wiki/Getting-Started:-Install-&-Use.

![screen-capture](https://cloud.githubusercontent.com/assets/5934455/25361023/2094cb96-2909-11e7-8b05-678a4701c2c4.JPG)

Thanks!

**Comment by patrick-rodgers** [comment]
Your code is executing before the polyfills have loaded. I just [updated the polyfill section of that article](https://github.com/SharePoint/PnP-JS-Core/wiki/Getting-Started:-Install-&-Use#option-1---polyfill-service) to describe how to use a service as well as a callback function to ensure things are loaded before your code executes. This should get you going.

**Comment by koltyakov** [comment]
I'd guess that after a page is loaded, there is no Promise object but ES6Promise.
As far as I know es6-promise.js has different versions, some do not define/polyfill Promise object with the same name.
I faced such or similar issue a couple of days ago but didn't have time for reasonable investigation and the issue was temporarily solved by placing `var Promise = Promise || ES6Promise;` after es-promise.js and before pnp related code.

**Comment by patrick-rodgers** [comment]
There are two versions of the file I think @koltyakov - one does the auto global fill and the other doesn't and expects you to use it in a module loader. I've started using the method I added to the wiki article as it is easier and was on my list to update that article. Which this prompted.

@astatke - any luck?

**Comment by koltyakov** [comment]
@patrick-rodgers, in my case it's an intranet without access to 3rd party services and websites.
`one does the auto global fill and the other doesn't` - yep, in my installation, I'll gonna using the one with global auto fill.

**Comment by patrick-rodgers** [comment]
Makes sense to me - I guess I can see why they have both, but seems confusing and causes problems like in this thread. When I first wrote those instructions I guess the one I used auto filled, and then later ones that behavior seemed to change.

**Comment by astatke** [comment]
Hi Patrick, I tried your "Option 1 - Polyfill Service" but now I am having error on Chrome Browser as well @patrick-rodgers @koltyakov. Have a look to the screenshot below. 
![error-on-chrome](https://cloud.githubusercontent.com/assets/5934455/25448457/ca8dbdf2-2a75-11e7-8667-bba6ca8bab69.JPG)

**Comment by patrick-rodgers** [comment]
Based on that I am guessing you are working against SP 2013 on-premises? If so you do not have the support for minimal metadata installed in your farm and need to do [one additional configuration as outlined here](https://github.com/SharePoint/PnP-JS-Core/wiki/Configuration#sharepoint-2013-on-premises).

**Comment by astatke** [comment]
@patrick-rodgers Apparently I am using SharePoint 2016 on-premise. 
On a side note I am curious as to how I can implement using Module Loader. Sorry, never done this before.
I also wonder if @koltyakov can share as to how you resolved your issue with Internet explorer.

**Comment by koltyakov** [comment]
Hi @astatke,

Please check the version of es6-promise you use.

With es6-promise 3.x.x my setup was:

```html
<script src="../es6-promise/es6-promise.min.js"></script>
<script src="../whatwg-fetch/fetch.min.js"></script>
<script src="../sp-pnp-js/pnp.min.js"></script>
```

or combined in the same sequence in a bundle.

With es6-promise 4.x.x they issued es6-promise and es6-promise-auto.
For es6-promise my initial quick workaround was adding:

```html
<script> var Promise = Promise || ES6Promise; </script>
```

right after `<script src="../es6-promise/es6-promise..minjs"></script>`.
For -auto version, there is no need to:

```html
<script src="../es6-promise/es6-promise-auto.min.js"></script>
<script src="../whatwg-fetch/fetch.min.js"></script>
<script src="../sp-pnp-js/pnp.min.js"></script>
```

on-premises tenants also require setup headers:

```javascript
$pnp.setup({
    headers: {
        'Accept': 'application/json; odata=verbose'
    }
});
```

before starting using the lib.

I'm constantly switching between, numerous SPO, 2013, 2016. The only surprise was with es6-promise without auto global fill.

**Comment by astatke** [comment]
Thank you so much @koltyakov and @patrick-rodgers. 

Great! Adding this line of code, `<script> var Promise = Promise || ES6Promise; </script>`, immediately after the this line of code `<script src="../es6-promise/es6-promise..minjs"></script>` helped me to display my content on IE browser. Although, I am on-premise SP 2016 environment I didn't have to add this code in my implementation:

`$pnp.setup({
    headers: {
        'Accept': 'application/json; odata=verbose'
    }
});`

On a side note, I wonder what you guys recommend as my plan was to use this implementation in a prod environment.

Once again thank you for your prompt response on this :1st_place_medal:

**Comment by patrick-rodgers** [comment]
@astatke - we recommend when possible deploying the library to a CDN for production use. How you actually package and deploy your code to production is a topic outside this library as there are many answers dependent on many factors.

Going to close this as answered as the original question appears resolved. Thanks!

**Comment by enitesh** [comment]
@koltyakov and @patrick-rodgers Thanks a lot for your suggestions. I was facing the same issue on Angular 5 application on IE 11. The ES6 error function was not working on IE and I have resolved it by following steps mentioned in link https://github.com/SharePoint/PnP-JS-Core/wiki/Getting-Started:-Install-&-Use. I have followed below steps

1. npm install es6-promise --save
2. npm install whatwg-fetch --save
3. In app.module.ts added below code 
import "zone.js/dist/zone";
import  "es6-promise";
import 'whatwg-fetch';

It's now working on IE11
